contracts/MechMarketplace.sol

```diff
@@ -0,0 +1,438 @@
+        }
+
+        // Record the actual delivery mech
+        mechDelivery.deliveryMech = msg.sender;
+
+        // Decrease the number of undelivered requests
+        numUndeliveredRequests--;
+
+        // Increase mech karma that delivers the request
+        mapMechKarma[msg.sender]++;
+
+        emit MarketplaceDeliver(priorityMech, msg.sender, account, requestId, requestData);
+
+        _locked = 1;
+    }
+
+    /// @dev Gets the already computed domain separator of recomputes one if the chain Id is different.
+    /// @return Original or recomputed domain separator.
+    function getDomainSeparator() public view returns (bytes32) {
+        return block.chainid == chainId ? domainSeparator : _computeDomainSeparator();
+    }
+
+    /// @dev Gets the request Id.
+    /// @param account Account address.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @return requestId Corresponding request Id.
+    function getRequestId(address account, bytes memory data) public pure returns (uint256 requestId) {
+        requestId = uint256(keccak256(abi.encode(account, data)));
+    }
+
+    /// @dev Gets the request Id with a specific nonce.
+    /// @param account Account address.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @param nonce Nonce.
+    /// @return requestId Corresponding request Id.
+    function getRequestIdWithNonce(
+        address account,
+        bytes memory data,
+        uint256 nonce
+    ) public view returns (uint256 requestId)
+    {
+        requestId = uint256(keccak256(
+            abi.encodePacked(
+                "\x19\x01",
+                getDomainSeparator(),
+                keccak256(
+                    abi.encode(
+                        account,
+                        data,
+                        nonce
+                    )
+                )
+            )
+        ));
+    }
+
+    /// @dev Gets the request Id status.
+    /// @param requestIdWithNonce Request Id with nonce.
+    /// @return status Request status.
+    function getRequestStatus(uint256 requestIdWithNonce) external view returns (RequestStatus status) {
+        // Request exists if it has a record in the mapRequestIdDeliveries
+        MechDelivery memory mechDelivery = mapRequestIdDeliveries[requestIdWithNonce];
+        if (mechDelivery.priorityMech != address(0)) {
+            // Check if the request Id was already delivered: delivery mech address is not zero
+            if (mechDelivery.deliveryMech == address(0)) {
+                status = RequestStatus.Requested;
+            } else {
+                status = RequestStatus.Delivered;
+            }
+        }
+    }
+
+    /// @dev Gets mech delivery info.
+    /// @param requestIdWithNonce Request Id with nonce.
+    /// @return Mech delivery info.
+    function getMechDeliveryInfo(uint256 requestIdWithNonce) external view returns (MechDelivery memory) {
+        return mapRequestIdDeliveries[requestIdWithNonce];
+    }
+}
+
```

Comment:
How does a non-priority mech easily monitor for expired requests and pick them up to work. What methods would it call to quickly and easily get that data>

Reply:
Oh we can actually add that into the `getRequestStatus()` function: request that is pending a priority mech execution, and expired request.

But as a general approach - go through all the mechs, browse each of their cyclic requests map - ask mech Marketplace about their status.

I'm thinking now we should push all the registered mech addresses in order to quickly retrieve them.

Reply:
@0xArdi is this good enough for you?

Reply:
@DavidMinarsch We synced on it, the way that is implemented right now is fine.